File: kingfisher-web/src/api/doc.ts
import request from './request';

export interface DocDirNode {
  id: number;
  parent_id: number;
  name: string;
  sort: number;
  status: number;
  /** 目录可见性 shared | private（private 仅 admin 可见，不公开） */
  visibility?: 'shared' | 'private';
  /** 该目录下当前用户可见的文档（目录树叶子节点） */
  docs?: Pick<DocItem, 'id' | 'title' | 'status' | 'visibility'>[];
  children?: DocDirNode[];
}

export interface DocItem {
  id: number;
  dir_id: number;
  title: string;
  content: string;
  owner_id: number;
  owner_name?: string;
  visibility: 'shared' | 'private';
  status: 'draft' | 'published';
  current_version: number;
  created_at: string;
  updated_at: string;
}

export interface DocVersion {
  id: number;
  doc_id: number;
  version_no: number;
  title: string;
  content: string;
  owner_name?: string;
  note: string;
  created_at: string;
}

/** 目录操作 */
export const docDirApi = {
  getTree: () => request.get('/docs/tree'),
  create: (data: { parent_id: number; name: string; sort?: number; visibility?: string }) =>
    request.post('/docs/dirs', data),
  update: (id: number, data: Record<string, unknown>) => request.put(`/docs/dirs/${id}`, data),
  delete: (id: number) => request.delete(`/docs/dirs/${id}`),
};

/** 公开文档（无需登录） */
export const publicDocApi = {
  get: (id: number) => request.get(`/public/docs/${id}`),
  getTree: () => request.get('/public/docs/tree'),
};

/** 文档操作 */
export const docApi = {
  upload: (file: File) => {
    const form = new FormData();
    form.append('file', file);
    return request.post('/docs/upload', form, { headers: { 'Content-Type': 'multipart/form-data' } });
  },
  list: (dir_id: number, params: Record<string, unknown>) => request.get('/docs', { params: { ...params, dir_id } }),
  getById: (id: number) => request.get(`/docs/${id}`),
  create: (data: { dir_id: number; title: string; content: string; visibility?: string; note?: string }) =>
    request.post('/docs', data),
  update: (id: number, data: { title: string; content: string; visibility?: string; note?: string }) =>
    request.put(`/docs/${id}`, data),
  publish: (id: number) => request.put(`/docs/${id}/publish`, {}),
  unpublish: (id: number) => request.put(`/docs/${id}/unpublish`, {}),
  versions: (id: number) => request.get(`/docs/${id}/versions`),
  getVersion: (id: number, no: number) => request.get(`/docs/${id}/versions/${no}`),
  restore: (id: number, version_no: number) => request.post(`/docs/${id}/restore`, { version_no }),
  delete: (id: number) => request.delete(`/docs/${id}`),
};

File: extends/doc/domain/doc.go
// Package doc implements document management logic.
package domain

import "time"

// 文档可见性
const (
	VisibilityShared  = "shared"  // 共享：目录可见者均可读
	VisibilityPrivate = "private" // 私有：仅作者（+admin）可读/改
)

// 文档生命周期状态
const (
	DocStatusDraft     = "draft"     // 草稿：仅作者/admin 可见
	DocStatusPublished = "published" // 已发布：按可见性规则公开
)

// DocDirectory 文档目录（树形）
type DocDirectory struct {
	ID           uint           `json:"id"`
	ParentID     uint           `json:"parent_id"`
	Name         string         `json:"name"`
	Sort         int            `json:"sort"`
	Status       int            `json:"status"`
	Version      string         `json:"version"`
	GrantedRoles []uint         `json:"granted_roles,omitempty"` // 可见角色 id 白名单（空 = 默认拒绝：仅 admin 可见）
	Docs         []DocTreeItem  `json:"docs,omitempty"`          // 该目录下当前用户可见的文档（叶子节点）
	Children     []DocDirectory `json:"children,omitempty"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
}

// DocTreeItem 目录树中的文档叶子节点（轻量，不含正文）
type DocTreeItem struct {
	ID         uint   `json:"id"`
	Title      string `json:"title"`
	Status     string `json:"status"`
	Visibility string `json:"visibility"`
}

// Document 文档
type Document struct {
	ID             uint       `json:"id"`
	DirID          uint       `json:"dir_id"`
	Title          string     `json:"title"`
	Content        string     `json:"content"` // Quill 输出的 HTML（当前最新内容）
	OwnerID        uint       `json:"owner_id"`
	OwnerName      string     `json:"owner_name,omitempty"`
	Visibility     string     `json:"visibility"`      // shared | private
	Status         string     `json:"status"`          // draft | published
	CurrentVersion int        `json:"current_version"` // 当前指向 doc_versions 的最新 version_no
	Sort           int        `json:"sort"`
	PublishedAt    *time.Time `json:"published_at,omitempty"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
}

// DocVersion 文档版本（append-only 历史）
type DocVersion struct {
	ID        uint      `json:"id"`
	DocID     uint      `json:"doc_id"`
	VersionNo int       `json:"version_no"`
	Title     string    `json:"title"`
	Content   string    `json:"content"`
	OwnerID   uint      `json:"owner_id"`
	OwnerName string    `json:"owner_name,omitempty"`
	Note      string    `json:"note"` // 变更说明
	CreatedAt time.Time `json:"created_at"`
}
